File: iOSWebBox/iOSWebBox/Components/EventComponent.swift
import Foundation
import CoconutSDK

/// Self-test topic emitted by the demo `echo` method.
public let EVENT_TOPIC_TEST_ECHO = "test.echo"

/**
 * Event Component
 *
 * Exposes subscribe/unsubscribe to H5 (delegating to the shared EventEmitter)
 * and a self-test `echo` method that emits a `test.echo` event after 500ms.
 */
public class EventComponent: BaseComponent {
    override public init() { super.init() }

    override public var name: String { "event" }
    override public var version: String { "1.0.0" }
    override public var pluginDescription: String { "Event subscription component" }

    private var sharedContext: ComponentContext?

    override public func onInit(context: ComponentContext) async {
        sharedContext = context
        Logger.shared.d(name, "Event component initialized")
    }

    override public func handle(function: String, params: [String: Any]?) async throws -> [String: Any] {
        switch function {
        case "subscribe": return try subscribe(params)
        case "unsubscribe": return try unsubscribe(params)
        case "echo": return try echo(params)
        default: try functionNotSupportedError(function)
        }
    }

    /// Register a subscription. Both topic and subscriptionId are H5-supplied.
    private func subscribe(_ params: [String: Any]?) throws -> [String: Any] {
        let topic = getParam(params, "topic")
        let subscriptionId = getParam(params, "subscriptionId")
        if topic.isEmpty || subscriptionId.isEmpty {
            try error(ErrorCode.PARAM_VALIDATION_FAILED, "topic and subscriptionId are required")
        }
        sharedContext?.eventEmitter.subscribe(topic: topic, subscriptionId: subscriptionId)
        return success(["subscriptionId": subscriptionId, "topic": topic])
    }

    private func unsubscribe(_ params: [String: Any]?) throws -> [String: Any] {
        let subscriptionId = getParam(params, "subscriptionId")
        if subscriptionId.isEmpty {
            try error(ErrorCode.PARAM_VALIDATION_FAILED, "subscriptionId is required")
        }
        sharedContext?.eventEmitter.unsubscribe(subscriptionId: subscriptionId)
        return success(["subscriptionId": subscriptionId, "success": true])
    }

    /// Demo: emit `test.echo` with the supplied payload after 500ms.
    private func echo(_ params: [String: Any]?) throws -> [String: Any] {
        let payload = params ?? [:]
        let topic = EVENT_TOPIC_TEST_ECHO
        Task { @MainActor [weak self] in
            try? await Task.sleep(nanoseconds: 500_000_000)
            self?.sharedContext?.eventEmitter.emit(topic: topic, data: payload)
            Logger.shared.d("event", "echo emitted: \(topic)")
        }
        return success(["scheduled": true, "topic": topic])
    }
}

File: AndroidWebBox/app/src/main/java/com/sniper/androidwebbox/components/event/EventComponent.kt
package com.sniper.androidwebbox.components.event

import com.sniper.coconut.component.BaseComponent
import com.sniper.coconut.component.ComponentContext
import com.sniper.coconut.component.ComponentMetadata
import com.sniper.coconut.utils.Logger
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/**
 * Event Component
 *
 * Exposes on/off to H5 (delegating to the shared EventEmitter) and a
 * self-test `echo` method that emits a `test.echo` event after 500ms.
 */
@ComponentMetadata(
    name = "event",
    version = "1.0.0",
    description = "Event subscription component",
    dependencies = []
)
class EventComponent : BaseComponent() {

    override val name = "event"
    override val version = "1.0.0"
    override val description = "Event subscription component"
    override val methods = listOf("on", "off", "echo")

    private lateinit var sharedContext: ComponentContext

    override suspend fun handle(function: String, params: JsonObject?): JsonElement {
        return when (function) {
            "on" -> onHandler(params)
            "off" -> offHandler(params)
            "echo" -> echo(params)
            else -> functionNotSupportedError(function)
        }
    }

    protected override suspend fun onInit(context: ComponentContext) {
        sharedContext = context
        Logger.d(name, "Event component initialized")
    }

    /**
     * Register a handler for `topic`. One handler per topic; second on overwrites.
     */
    private suspend fun onHandler(params: JsonObject?): JsonElement {
        val topic = getParam(params, "topic")

        if (topic.isEmpty()) {
            return paramValidationError("topic is required")
        }

        sharedContext.eventEmitter.on(topic)

        return buildJsonObject {
            put("topic", JsonPrimitive(topic))
        }.let { success(it) }
    }

    private suspend fun offHandler(params: JsonObject?): JsonElement {
        val topic = getParam(params, "topic")
        if (topic.isEmpty()) {
            return paramValidationError("topic is required")
        }
        sharedContext.eventEmitter.off(topic)

        return buildJsonObject {
            put("topic", JsonPrimitive(topic))
            put("success", JsonPrimitive(true))
        }.let { success(it) }
    }

    /**
     * Demo: emit `test.echo` with the supplied payload after 500ms.
     */
    private suspend fun echo(params: JsonObject?): JsonElement {
        val payload: JsonElement = params ?: buildJsonObject {}
        scope.launch {
            delay(500)
            sharedContext.eventEmitter.emit(TOPIC_TEST_ECHO, payload)
            Logger.d(name, "echo emitted: $TOPIC_TEST_ECHO")
        }

        return buildJsonObject {
            put("scheduled", JsonPrimitive(true))
            put("topic", JsonPrimitive(TOPIC_TEST_ECHO))
        }.let { success(it) }
    }

    companion object {
        const val TOPIC_TEST_ECHO = "test.echo"
    }
}
